week08/manager.go

```diff
@@ -0,0 +1,191 @@
+package week08
+
+import (
+	"context"
+	"sync"
+)
+
+// Manager is responsible for receiving product orders
+// and assigning them to employees. Manager is also responsible
+// for receiving completed products, and listening for errors,
+// from employees. Manager takes products that have been built
+// by employees and returns them to the customer as a CompletedProduct.
+type Manager struct {
+	Warehouse *Warehouse
+	cancel    context.CancelFunc
+	completed chan CompletedProduct
+	errs      chan error
+	jobs      chan *Product
+	stopped   bool
+
+	sync.RWMutex
+}
+
+// Start will create new employees for the given count,
+// and start listening for jobs and errors.
+// Managers should be stopped using the Stop method
+// when they are no longer needed.
+func (m *Manager) Start(ctx context.Context, count int) (context.Context, error) {
+
+	if count <= 0 {
+		return nil, ErrInvalidEmployeeCount(count)
+	}
+
+	// create a new cancellation context
+	ctx, cancel := context.WithCancel(ctx)
+	m.cancel = cancel
+
+	if m.Warehouse == nil {
+		m.Warehouse = &Warehouse{}
+	}
+
+	// start the warehouse
+	// this returns a context that can be listened to
+	// for cancellation notification from the warehouse
+	ctx = m.Warehouse.Start(ctx)
+
+	// launch a goroutine to listen context cancellation
+	go func() {
+		// listen for context cancellation
+		// this could come from the external context
+		// passed to m.Start()
+		<-ctx.Done()
+
+		// call the cancel function
+		cancel()
+
+		// call Stop()
+		m.Stop()
+	}()
```

Comment:
As we saw in week 8's discussion group, we discovered that there is a major race condition in this block that needs to be addressed.